drivers: can: Rework CAN drivers and API

alexanderwachter opened the pull request:
This PR extends the isr callback and the attach_isr function with an argument.
With this argument it's possible to map different filters to the same isr but having the possibility to distinguish them by the argument. This makes the implementation if higher level protocols easier but breaks the current API.
The sram overhead is sizeof(void*) * CAN_MAX_FILTER

EDIT:
attach_msgq is moved to a common function to all drivers. This reduces the complexity of the specific implementations.

attach_workq is implemented. This is useful to offload work from an ISR to a workq thread.
This function can be used to attach a callback even from userspace.
This is not possible for attach_isr.

Commit:
drivers: can: Extend attach_isr with void *arg and pass to the isr cb

This commit extends the CAN api attach_isr function by an argument
that is passed to the isr callback.
With this extension it is possible to distinguish between filter matches
when they use the same callback.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented:
@lowlander what do you think?

alexanderwachter commented:
@karstenkoenig ping

jukkar reviewed (approved):
This makes lot of sense so +1

lowlander commented:
My old driver that I have almost cleaned up enough to create a PR for, has a bit of another approach. I give my driver a (const) array of filters, that are directly used by the driver, so I do not copy them to not need writable RAM (or a fixed CONFIG_FILTER_COUNT option). Also I made it possible to set a filter-array per FIFO (there are two RX fifos on the F4) Also I have no "detach" system-call because just registering a new array (which can be 0) releases the previous one.  Those array entries, have not just the filter and callback, but also a user-pointer, that will be passed to the callback. 

Currently I am implementing a E70 CAN-FD driver, to make sure the API's work on more than just a STM32 controller.   

I am a bit afraid things are done in parallel at the moment, but last time I early asked for comments I was pretty much shot down due to code quality, so now I just finish my driver and create a PR when I am done.

jukkar commented:
Regarding the API break that this PR is introducing, should we just keep the old API and create a new function that introduces the extra arg to the callback. The old API could be deprecated if needed and eventually removed at some later time.

alexanderwachter commented:
@jukkar I'm not sure if the attach isr is used by users anyway. From userspace it's better to use the queue. The attach_isr is more useful for highlevel stuff. But making a wrapper will not produce any overhead. I'll do that.

alexanderwachter commented:
> I give my driver a (const) array of filters, that are directly used by the driver, so I do not copy them to not need writable RAM (or a fixed CONFIG_FILTER_COUNT option)
I don't copy the filters neither. I just set the registers.  CONFIG_FILTER_COUNT is for the amount of different callbacks/message queues and they can't be constant because socket can relies on appending filter at run-time.
Also my ISO-TP stack I'm working on relies in attach/detach at run-time. Additionally an other project (the project I implement all the CAN stuff here) relies on attach and detach at run-time.

Commit:
drivers: can: Extend can_send with void *arg and pass it to the isr cb

This commit extends the CAN api can_send function by an argument
that is passed to the isr callback.
With this extension it is possible to distinguish between sent masseges
when they use the same callback.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented:
Added an argument to the send callback too.

alexanderwachter commented:
Rebased.

Commit:
drivers: can: rework can_attach_msgq

can_attach_msgq can be implemented as a wrapper of can_attach_isr.
This is implemented as a common function for all drives and reduces
the complexity of the specific drivers. Since this is common to
multi instances of drivers too, it is removed from the API struct.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
drivers: can: Implement can_attach_workq

can_attach_workq is an isr wrapper that puts a work item into a workq
whenever a received frame matches the filter. With this function it is
possible to have a callback that is offloaded. This is useful if the
work is too complex for an isr or USERSPACE is enabled.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
tests: drivers: can: api: define DT_CAN_1_NAME if not defined

Define DT_CAN_1_NAME as an empty string if it is not defined.
This is useful for buildtest where no HW is available.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented:
can_attach_workq is implemented as a replacement for can_attach_isr when userspace is on.
attaching an ISR from userspace is not allowed.(See #14204)

Commit:
tests: drivers: can: Add testcases for attach_workq

Add testcases for the new attach_workq API.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

jukkar reviewed (commented):
One possible issue with the API changes is that we are changing the API here. There is the doc for API mgmt https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html. For 1.14 lot of API changes were done that did not follow the rules verbatim, but it looks like that for 1.15 the rules are more strictly followed.
Anyway, as I am not really using the CAN low level APIs, perhaps those one that do could comment these changes.

Commit:
tests: drivers: can: Make CAN dev global

Get the device binding only once and write it to a global variable.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
drivers: can: Add loopback driver

This commit implements a CAN loopback device. This device is used
for testing when no CAN controller is available on the hardware.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented:
@karstenkoenig can you check the changes for mcp2515?

rodrigopex commented on drivers/can/can_loopback.c:
```diff
@@ -0,0 +1,225 @@
+		}
+	}
+
+	return -1;
```
return -EIO...

rodrigopex commented on drivers/can/can_loopback.c:
```diff
@@ -0,0 +1,225 @@
+	if (filter_id < 0) {
+		LOG_ERR("No free filter left");
+		k_mutex_unlock(&data->mtx);
+		return CAN_NO_FREE_FILTER;
```
I would suggest using a standard error number. Something like -EBUSY. I am guessing there is no negative filter ID right?

alexanderwachter replied:
Right. But the CAN error numbers are defined along with the API to get a more precise error message.

rodrigopex commented on drivers/can/can_loopback.c:
```diff
@@ -0,0 +1,225 @@
+{
+	struct can_loopback_data *data = DEV_DATA(dev);
+
+	ARG_UNUSED(bitrate);
```
I usually put the unused variables at the first line of the function implementation. First the unused and after that, the code. I don't know if it is a good practice or just my of doing that.

alexanderwachter replied:
I thought the definitions have to be at the beginning.

rodrigopex commented on drivers/can/can_loopback.c:
```diff
@@ -0,0 +1,225 @@
+static int can_loopback_init(struct device *dev)
+{
+	struct can_loopback_data *data = DEV_DATA(dev);
+	int i;
```
The same about the scope of the **i**. I always try to shrink that.

rodrigopex commented on drivers/can/can_loopback.c:
```diff
@@ -0,0 +1,225 @@
+#ifdef CONFIG_CAN_1
+
+static const struct can_loopback_config can_loopback_cfg_1 = {
+};
```
Did you mean **{0}**?

alexanderwachter replied:
can_loopback_cfg_1 is not used at the moment.

alexanderwachter replied:
Removed it.

rodrigopex commented on samples/drivers/CAN/src/main.c:
```diff
@@ -172,10 +176,12 @@ void led_thread(void *msgq, void *can_dev_param, void *gpio_dev_param)
-void rx_button_isr(struct zcan_frame *msg)
+void rx_button_isr(struct zcan_frame *msg, void *arg)
 {
 	u16_t cnt = msg->data[0] | (msg->data[1] << 8);
```
I usually change the position to ```u16_t cnt = (msg->data[1] << 8) | msg->data[0];``` It seems to be more natural.

alexanderwachter replied:
The sample needs to be updated anyways. I'll provide a new one after this gets merged.

rodrigopex commented on samples/drivers/CAN/src/main.c:
```diff
@@ -172,10 +176,12 @@ void led_thread(void *msgq, void *can_dev_param, void *gpio_dev_param)
 {
 	u16_t cnt = msg->data[0] | (msg->data[1] << 8);
 
+	ARG_UNUSED(arg);
```
Same about UNUSED first...

rodrigopex reviewed (commented):
Here are my 2 cents. I hope I could help here.

alexanderwachter commented on drivers/can/can_loopback.c:
```diff
@@ -0,0 +1,225 @@
+		      void *callback_arg)
+{
+	struct can_loopback_data *data = DEV_DATA(dev);
+	int i;
```
Right. I thought it is now allowed. I'll change it.

alexanderwachter commented on drivers/can/CMakeLists.txt:
```diff
@@ -1,5 +1,7 @@
+zephyr_sources_ifdef(CONFIG_CAN          can_common.c)
+zephyr_sources_ifdef(CONFIG_CAN_LOOPBACK can_loopback.c)
 zephyr_sources_ifdef(CONFIG_CAN_MCP2515  mcp2515.c)
+zephyr_sources_ifdef(CONFIG_CAN_STM32    stm32_can.c)
```
That's right. I can make a commit for that.

alexanderwachter replied:
Done

alexanderwachter commented on drivers/can/can_common.c:
```diff
@@ -0,0 +1,145 @@
+	u16_t next_head = WORK_BUF_MOD_SIZE(buffer->head + 1);
+
+	if (buffer->head == WORK_BUF_FULL) {
+		return -1;
```
Good point. I could also use void and throw the error log here.

alexanderwachter replied:
-EIO seems not really to be ideal here because it is not an IO error. I can't find a suitable error number like buffer full. Since this is defined static and only used at one location I leave it as -1 here

Commit:
drivers: can: rename drivers to can_<dev name>

The filenames of the can drivers were not consisten.
Changed them to can_<dev name>

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented on drivers/can/can_handlers.c:
```diff
@@ -44,6 +47,25 @@ Z_SYSCALL_HANDLER(can_attach_msgq, dev, msgq, filter) {
+				     sizeof(struct zcan_filter)));
+	Z_OOPS(Z_SYSCALL_MEMORY_READ((void *)callback_arg, sizeof(void *)));
+
+	Z_OOPS(Z_SYSCALL_OBJ(work_q, K_OBJ_MSGQ));
```
This is obviously wrong. But which K_OBJ_ is the right one here? I can't find K_OBJ_WORKQ @andrewboie

andrewboie replied:
a k_work_q is just a k_queue and a k_thread. we don't do anything to the thread, we are just going to be adding items to the k_queue object. so it's enough to do Z_OOPS(Z_SYSCALL_OBJ(&work_q->queue, K_OBJ_QUEUE).

andrewboie commented on drivers/can/can_handlers.c:
```diff
@@ -47,6 +47,25 @@ Z_SYSCALL_HANDLER(can_attach_msgq, dev, msgq, filter) {
+
+	return z_impl_can_attach_workq((struct device *)dev,
+				     (struct k_work_q *)work_q,
+				     (struct zcan_work *)work,
```
No validation has been done to the work parameter. Have to prove that the caller has read/write access to it.

andrewboie commented on drivers/can/can_common.c:
We have a security problem here.
The pointer to the zcan_work came from user memory. The problem is we will be using this data structure when the ISR fires. But what is going to stop the user thread from concurrently overwriting the members of the zcan_work struct with garbage or malicious data? If that happens, terrible things might happen can_work_isr_put() actually runs depending on what was changed, particularly the work->work_queue pointer.

alexanderwachter replied:
@andrewboie I could check the data within the ISR. The user can't run concurrently to an ISR.

alexanderwachter replied:
An alternative would be making the zcan_work a kernel object. But how can I copy the zcan_frame to the userspace then?

alexanderwachter replied:
@andrewboie Userspace handler for attaching workqueues is removed until workqueues get proper userspace support.

andrewboie replied:
since there is no handler can you remove the __syscall

alexanderwachter replied:
Removed.

tyrantlizard77 reviewed (dismissed):
I'm seeing some problems with the workq integration with respect to user mode.
Doing this right may require some adjustments to the workq interfaces, I will follow up with some more thoughts on this.

alexanderwachter commented on drivers/can/can_common.c:
How do I know if the work_q thread is in user mode?

karstenkoenig reviewed (approved):
Looks all good to me, glad to see attachmsqq gone from the hardware drivers!
Also still works on my setup here :-)

alexanderwachter commented on drivers/can/Kconfig.loopback:
```diff
@@ -0,0 +1,44 @@
+# Kconfig.stm32 - STM32 CAN configuration options
```
Thanks! Resolved

alexanderwachter commented on drivers/can/can_mcp2515.c:
```diff
@@ -601,8 +584,7 @@ static int mcp2515_init(struct device *dev)
-		     sizeof(dev_data->filter_response));
+	(void)memset(dev_data->rx_cb, 0, sizeof(dev_data->rx_cb));
 	(void)memset(dev_data->filter, 0, sizeof(dev_data->filter));
 
```
They don't need to be initialized. They are not checked or used before an ISR is actually attached.
